### PR TITLE
Enable support for multiple platforms in the Linux Bridge CNI.

### PR DESCRIPTION
**Pull request by ashokpariya0**
These updates allow the building and pushing of Linux Bridge container images for multiple platforms (e.g., amd64, s390x) using a single Dockerfile. Multi-platform build support is provided for both Docker and Podman container runtimes.

**What this PR does / why we need it**:

"The upstream **cni-default-plugins** image ([[link](https://quay.io/repository/kubevirt/cni-default-plugins?tab=tags&tag=latest)](https://quay.io/repository/kubevirt/cni-default-plugins?tab=tags&tag=latest)) currently supports only the amd64 architecture. The proposed changes in this PR enable multi-platform support for the **linux-bridge** image, broadening its compatibility to include additional architectures like s390x."

**Special notes for your reviewer**:
Multi-platform build support is enabled for both Docker and Podman container runtimes, and the necessary changes have been made accordingly.

To use this feature, you need to:
For Podman:
No additional dependencies are required to build multi-platform images. Podman supports this feature natively.
and We set the default value of OCI_BIN to podman if it's available, so we are good here.

For Docker:
Have access to docker buildx. More information: [[Docker Buildx Documentation](https://docs.docker.com/build/building/multi-platform/)](https://docs.docker.com/build/building/multi-platform/)
Ensure that BuildKit is enabled. More information: [[Docker BuildKit Enhancements](https://docs.docker.com/develop/develop-images/build_enhancements/)](https://docs.docker.com/develop/develop-images/build_enhancements/)

**Release note**:

```release-note
None
```

### CODE REVIEW

File: hack/components/bump-linux-bridge.sh
Comment (oshoval) [inline hack/components/bump-linux-bridge.sh:31]: please add auto detection instead hardcoding both, we should be able to compile as fast as possible

File: hack/components/bump-linux-bridge.sh
Comment (oshoval) [inline hack/components/bump-linux-bridge.sh:49]: please drop AS final, it is not used
prone theoretically for mistakes if someone adds layers below that

File: hack/components/bump-linux-bridge.sh
Comment (oshoval) [inline hack/components/bump-linux-bridge.sh:83]: why no-cache ?
we do want cache locally (fast build is important)

File: hack/components/bump-linux-bridge.sh
Comment (oshoval) [inline hack/components/bump-linux-bridge.sh:93]: i dont understand why we sed -i instead directly building the dockerfile as needed please
after all the file is built in `create_dockerfile` right?
lets refactor this please

File: hack/components/bump-linux-bridge.sh
Comment (oshoval) [inline hack/components/bump-linux-bridge.sh:117]: please move it after `build_podman_image`

**Review by oshoval — COMMENTED** [review]
Thanks, some comments

File: hack/components/bump-linux-bridge.sh
Comment (ashokpariya0) [reply hack/components/bump-linux-bridge.sh]: Done.

File: hack/components/bump-linux-bridge.sh
Comment (ashokpariya0) [reply hack/components/bump-linux-bridge.sh]: Done.

File: hack/components/bump-linux-bridge.sh
Comment (ashokpariya0) [reply hack/components/bump-linux-bridge.sh]: As discussed no-cache is removed to speed up build.

File: hack/components/bump-linux-bridge.sh
Comment (ashokpariya0) [reply hack/components/bump-linux-bridge.sh]: Done

File: hack/components/bump-linux-bridge.sh
Comment (ashokpariya0) [reply hack/components/bump-linux-bridge.sh]: We need to use `TARGETARCH` and `TARGETOS` in the Dockerfile, but these values are only available at runtime. Since the Dockerfile is generated using a `cat` command, directly inserting these variables during creation would result in empty values, leading to an incorrect Dockerfile. To address this, we generate the Dockerfile first and then update it using the `sed` command to properly set the values at runtime.

Comment (oshoval) [reply hack/components/bump-linux-bridge.sh]: Just out of curiosity please (no need to do it, because it means we need a new file)

Cant we do the same you did before, have those in a new actual Dockerfile (not one that is catted),
and supply them externally, exactly as in other places ?
```
ARG TARGETOS
ARG TARGETARCH
```

File: hack/components/bump-linux-bridge.sh
Comment (oshoval) [inline hack/components/bump-linux-bridge.sh:37]: please use default value bash syntax, as in line 30

**Review by oshoval — COMMENTED** [review]
Just a small comment, thanks

File: hack/components/bump-linux-bridge.sh
Comment (oshoval) [reply hack/components/bump-linux-bridge.sh]: I am a bit confused why it is different than all the other cases where we have a real Dockerfile (not catted)
with those as args, but we can take it offline, and unblock this PR, (because in any case, not sure we want a standalone Dockerfile for the bump / publish)

File: hack/components/bump-linux-bridge.sh
Comment (ashokpariya0) [reply hack/components/bump-linux-bridge.sh]: Done.

File: hack/components/bump-linux-bridge.sh
Comment (ashokpariya0) [reply hack/components/bump-linux-bridge.sh]: Fixed this, Removed sed commands and used backslash to avoid substitution while creating file for few specific lines with cat command.

**Comment by oshoval** [comment]
Looks great thanks

/test all
/approve

@0xFelix 
can you please take a look ?

**Review by 0xFelix — COMMENTED** [review]
Can you run `shellcheck` on this file and fix the issues introduced by this PR?

**Comment by oshoval** [comment]
> Can you run `shellcheck` on this file and fix the issues introduced by this PR?

Note that some the warnings are considered annoyance imho, so it is better to have judgement per warning please
(i even saw some delicate ones that can create bugs, but on more sophisticated scripts like we have on kubevirtci sriov)

**Comment by ashokpariya0** [comment]
> Can you run `shellcheck` on this file and fix the issues introduced by this PR?
>
This PR does not introduce any new warnings; I have already verified it using shellcheck

**Comment by oshoval** [comment]
Once this PR is merged, also need to recall / find out please how to trigger its release

**Comment by ashokpariya0** [comment]
> ```
> In /home/felix/Downloads/bump-linux-bridge.sh line 95:
>         if [ ! -z "${PUSH_IMAGES}" ]; then
>              ^-- SC2236 (style): Use -n instead of ! -z.
> 
> 
> In /home/felix/Downloads/bump-linux-bridge.sh line 103:
>     cd ${LINUX_BRIDGE_PATH}
>        ^------------------^ SC2086 (info): Double quote to prevent globbing and word splitting.
> 
> Did you mean:
>     cd "${LINUX_BRIDGE_PATH}"
> ```
> 
> These are relevant.
>



> The 2nd isnt regression of this PR, as long as we don't afraid from word splitting / globbing in our values i prefer to not to enforce it please.
> 
> I remember something vaguely about -n when not using quotes, if the value is empty, it can have undesired output, but value shouldn't be empty here. If needed I can find it.
> 
> More over, the 1st isn't regression of this PR, so imo we don't need to enforce fixing it on this PR (and it is syntactic equivalent)
>
@oshoval I also encountered some issues recently when trying to achieve a 100% result in ShellCheck without any warnings.
Since none of the warnings above were introduced by this PR, can we proceed with it? I can submit a follow-up PR if needed.  @0xFelix

**Comment by oshoval** [comment]
imo we can proceed please,
and even we don't need a follow-up about it from my side
thanks

**Comment by 0xFelix** [comment]
I'm not a fan of leaving warnings in place when you change the code anyway. But it up to you to decide.

**Comment by ashokpariya0** [comment]
@0xFelix @oshoval  Is there anything else on your end? If not, could we proceed with merging this PR?

**Comment by ashokpariya0** [comment]
> hack/components/bump-linux-bridge.sh

> btw once this PR is merged Looking on #1767 I think you can just change to "tagged" and the auto bump will do the rest since there is a new tag https://github.com/containernetworking/plugins/releases/tag/v1.6.1 Best if you can verify please with Or Mergi this is correct
> 
I created issue https://github.com/kubevirt/cluster-network-addons-operator/issues/1937 to discuss above.
> you will need to include `PLATFORMS=all` of course in that flow somehow maybe you can set it as default and thats it for simplicity, in case of problem we can rethink
>
Do you mean that we need to set the default PLATFORMS as all in the script hack/components/bump-linux-bridge.sh
?

**Comment by oshoval** [comment]
@0xFelix can you please re-review ?

**Comment by oshoval** [comment]
> I created issue #1937 to discuss above.
> 
> > you will need to include `PLATFORMS=all` of course in that flow somehow maybe you can set it as default and thats it for simplicity, in case of problem we can rethink
> 
> Do you mean that we need to set the default PLATFORMS as all in the script hack/components/bump-linux-bridge.sh ?

Thanks
Either that, or the script need to be smart that if it runs on via git actions it enables it, would you able please to check about it ? if it is complex, we can just set it hardcoded as long as it can still run locally easily

**Comment by ashokpariya0** [comment]
> > > you will need to include `PLATFORMS=all` of course in that flow somehow maybe you can set it as default and thats it for simplicity, in case of problem we can rethink
> > 
> > 
> > Do you mean that we need to set the default PLATFORMS as all in the script hack/components/bump-linux-bridge.sh ?
> 
> Thanks Either that, or the script need to be smart that if it runs on via git actions it enables it, would you able please to check about it ? if it is complex, we can just set it hardcoded as long as it can still run locally easily
>
Let me add this change to the PR itself. I don’t foresee any issues with running it locally, as we use Podman, and there are no additional dependencies required on Podman for the multi-platform image build.

**Comment by oshoval** [comment]
ok, just to make sure, it doesnt break docker right?
we still do support docker unless we decide in the future to stop supporting it

**Comment by ashokpariya0** [comment]
> ok, just to make sure, it doesnt break docker right? we still do support docker unless we decide in the future to stop supporting it
>
Correct, it will work with Docker as well. The only requirement for Docker is the use of docker buildx, while no additional dependencies are needed for Podman.

**Comment by oshoval** [comment]
lgtm

**Comment by oshoval** [comment]
This PR must be merged before the one that update the tag does, else it will be harder to trigger the auto bumper

**Comment by ashokpariya0** [comment]
cc @phoracek

**Comment by oshoval** [comment]
We need Felix review, no need to tag here Petr atm please

**Comment by ashokpariya0** [comment]
@oshoval @0xFelix I added a Podman command to remove dangling images. Please refer to the explanation here:
https://github.com/k8snetworkplumbingwg/kubemacpool/pull/444#discussion_r1883313463

**Comment by oshoval** [comment]
Thanks

if the dangling flag is crucial we might need to add it on all the other repos ?
anyhow no rush, in case there were more nits so we can do them at once

/test all

**Comment by oshoval** [comment]
https://github.com/kubevirt/cluster-network-addons-operator/actions/runs/12310655279/job/34359580353?pr=1927
not related to this PR, seems kubevirt deployment didnt finish in time, first time i see it, so no biggie for now

**Comment by oshoval** [comment]
@RamLavi can you please review ?

**Review by RamLavi — APPROVED** [review]
/lgtm

**Comment by oshoval** [comment]
something is broken since latest pushes, please check, you can create a vanilla PR to see main does work if desired

**Comment by oshoval** [comment]
/hold

so it wont be retested

**Comment by ashokpariya0** [comment]
> something is broken since latest pushes, please check, you can create a vanilla PR to see main does work if desired

> I think PR: https://github.com/kubevirt/cluster-network-addons-operator/pull/1943 got merged 3hrs back, May be i need to rebase my pr

**Comment by ashokpariya0** [comment]
> something is broken since latest pushes, please check, you can create a vanilla PR to see main does work if desired
>
 I think PR: #1943 got merged 3hrs back, May be i need to rebase my pr

**Comment by ashokpariya0** [comment]
> > something is broken since latest pushes, please check, you can create a vanilla PR to see main does work if desired
> 
> I think PR: #1943 got merged 3hrs back, May be i need to rebase my pr
>
Done

**Comment by oshoval** [comment]
seems https://github.com/kubevirt/cluster-network-addons-operator/actions/runs/12339360372/job/34435568203?pr=1927 is failing, would you have time to look please ?
it seems it is just because the packages it tries to uninstall arent there,
simple fix it to `|| true` with that
better is to check what https://github.com/ovn-kubernetes/ovn-kubernetes are doing on their lane, because they were the ref

**Comment by ashokpariya0** [comment]
> https://github.com/ovn-kubernetes/ovn-kubernetes



> Thanks
> 
> seems https://github.com/kubevirt/cluster-network-addons-operator/actions/runs/12339360372/job/34435568203?pr=1927 is failing, would you have time to look please ? it seems it is just because the packages it tries to uninstall arent there, simple fix it to `|| true` with that better is to check what https://github.com/ovn-kubernetes/ovn-kubernetes are doing on their lane, because they were the ref
>
I don't see any change w.r.t to free disk space https://github.com/ovn-kubernetes/ovn-kubernetes/blob/eee91422cc619cde68dcb4e3be30ccf33e188867/.github/workflows/test.yml#L295

**Comment by oshoval** [comment]
> > https://github.com/ovn-kubernetes/ovn-kubernetes
> 
> > Thanks
> > seems https://github.com/kubevirt/cluster-network-addons-operator/actions/runs/12339360372/job/34435568203?pr=1927 is failing, would you have time to look please ? it seems it is just because the packages it tries to uninstall arent there, simple fix it to `|| true` with that better is to check what https://github.com/ovn-kubernetes/ovn-kubernetes are doing on their lane, because they were the ref
> 
> I don't see any change w.r.t to free disk space https://github.com/ovn-kubernetes/ovn-kubernetes/blob/eee91422cc619cde68dcb4e3be30ccf33e188867/.github/workflows/test.yml#L295

i believe git actions default container was changed and it will fail there soon as well if we use the same

**Comment by oshoval** [comment]
unit test fail on kmp image mismatch (might be related to the last PR that was merged)
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_cluster-network-addons-operator/1927/pull-cluster-network-addons-operator-unit-test/1868297019211124736

maybe @RamLavi  knows something about it ?
```
diff --git a/test/releases/99.0.0.go b/test/releases/99.0.0.go
index b3c8ca32..2784c3fa 100644
--- a/test/releases/99.0.0.go
+++ b/test/releases/99.0.0.go
@@ -42,7 +42,7 @@ func init() {
 				ParentName: "kubemacpool-mac-controller-manager",
 				ParentKind: "Deployment",
 				Name:       "manager",
-				Image:      "quay.io/kubevirt/kubemacpool@sha256:eebb65b8a12cbfc20a429bbba399eb5a5c2279f8613c36965957ee7c36cfcbd6",
+				Image:      "quay.io/kubevirt/kubemacpool@sha256:15c7be1e7aad2836237e9a73443192bfa58c044fcc5488b175222f3ae997064f",
 			},
 			{
 				ParentName: "kubemacpool-mac-controller-manager",
@@ -54,7 +54,7 @@ func init() {
 				ParentName: "kubemacpool-cert-manager",
 				ParentKind: "Deployment",
 				Name:       "manager",
-				Image:      "quay.io/kubevirt/kubemacpool@sha256:eebb65b8a12cbfc20a429bbba399eb5a5c2279f8613c36965957ee7c36cfcbd6",
+				Image:      "quay.io/kubevirt/kubemacpool@sha256:15c7be1e7aad2836237e9a73443192bfa58c044fcc5488b175222f3ae997064f",
 			},
 			{
 				ParentName: "ovs-cni-amd64",
+ exit 1
make: *** [Makefile:101: check] Error 1
```

**Comment by oshoval** [comment]
open https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4914 and the one that is mentioned there
ipam-ext one (which has the same code as here) also fails for the same
lets see if OVN as well

**Comment by oshoval** [comment]
still fails after rebase, unit test should be fixed first, and worth to check if master state is broken 
after the KMP bump

**Comment by oshoval** [comment]
opened for that https://github.com/kubevirt/cluster-network-addons-operator/pull/1944

**Comment by oshoval** [comment]
storage problem also hit OVN
https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4914

**Comment by ashokpariya0** [comment]
> opened [ovn-kubernetes/ovn-kubernetes#4914](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4914) and the one that is mentioned there ipam-ext one (which has the same code as here) also fails for the same lets see if OVN as well
>
The issue also occurs with OVN, as it seems that ubuntu-latest now points to Ubuntu 24.04. This is causing disk cleanup problems across the board. We can address this by either removing only the installed packages or switching to ubuntu-22.04.

Additionally, the following link still doesn't reflect the correct version:

https://github.com/actions/runner-images

Please also review the issue here: https://github.com/actions/runner-images/issues/10636.

**Comment by oshoval** [comment]
Thank you
best to sync please with OVN people with your findings on https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4914 and choose the right direction
I will post that PR on slack once you post your findings there

**Comment by ashokpariya0** [comment]
> Thank you best to sync please with OVN people with your findings on [ovn-kubernetes/ovn-kubernetes#4914](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4914) and choose the right direction I will post that PR on slack once you post your findings there (or can just link your message, what ever you prefer)
>
Sure

**Comment by jommeke22f** [comment]
> Deze updates maken het mogelijk om Linux Zeecontainers te bouwen en te pushen voor meerdere platformen (bijv. amd64, s390x) met behulp van één Docker file. Ondersteuning voor multiplatform builds wordt geboden voor zowel Docker- als Persconferenties.
> 
> **Wat deze PR doet / waarom we het nodig hebben** :
> 
> "De upstream **calculations** image ([ [link](https://quay.io/repository/kubevirt/cni-default-plugins?tab=tags&tag=latest) ]( https://quay.io/repository/kubevirt/cni-default-plugins?tab=tags&tag=latest )) ondersteunt momenteel alleen de architecture. De voorgestelde wijzigingen in deze PR maken cultuurverandering mogelijk voor de **linux-bridge** image, waardoor de compatibiliteit wordt uitgebreid met extra architecturen zoals s390x."
> 
> **Speciale opmerkingen voor uw reviewer** : Ondersteuning voor multi-platform builds is ingeschakeld voor zowel Docker- als Postmodernisme en de nodige wijzigingen zijn dienovereenkomstig doorgevoerd.
> 
> Om deze functie te gebruiken, moet u het volgende doen: Voor Polman: Er zijn geen extra afhankelijkheden vereist om multi-platform images te bouwen. Podia ondersteunt deze functie standaard. en We stellen de standaardwaarde van OCI_BIN in op podium als deze beschikbaar is, dus we zitten hier goed.
> 
> Voor Docker: Heb toegang tot docker build. Meer informatie: [ [Docker Documentatie](https://docs.docker.com/build/building/multi-platform/) ]( https://docs.docker.com/build/building/multi-platform/ ) Zorg ervoor dat Builds is ingeschakeld. Meer informatie: [ [Docker Kwaliteitsverbetering](https://docs.docker.com/develop/develop-images/build_enhancements/) ]( https://docs.docker.com/develop/develop-images/build_enhancements/ )
> 
> **Release-opmerking** :
> 
> ```
> None
> ```

**Comment by ashokpariya0** [comment]
Before merging this PR, we need to first merge the following:

1. The PR that fixes the unit test issue with the Kubemacpool image: [#1946](https://github.com/kubevirt/cluster-network-addons-operator/pull/1946)
2. The PR that provides a workaround for the cleanup issue with the latest Ubuntu version: [#1945](https://github.com/kubevirt/cluster-network-addons-operator/pull/1945)

**Comment by oshoval** [comment]
can you please rebase to take ipam fix ?

**Comment by ashokpariya0** [comment]
> can you please rebase to take ipam fix ? we will also run the rest
>
I am on it.

**Comment by ashokpariya0** [comment]
> can you please rebase to take ipam fix ? we will also run the rest
>
Done.

**Comment by ashokpariya0** [comment]
@oshoval IPAM controller e2e tests passed on this pr, could you please trigger all other tests?

**Pull request merged**
